src/samplers/remote_sampler.js

```diff
@@ -0,0 +1,126 @@
+// @flow
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import request from 'request';
+import ProbabilisticSampler from './probabilistic_sampler.js';
+import RateLimitingSampler from './ratelimiting_sampler.js';
+import NullLogger from '../logger.js';
+
+const DEFAULT_INITIAL_SAMPLING_RATE = 0.001;
+const SAMPLING_REFRESH_INTERVAL = 60000;
+const DEFAULT_SAMPLING_PORT = 5778;
+const PROBABILISTIC_STRATEGY_TYPE = 0;
+const RATELIMITING_STRATEGY_TYPE = 1;
+
+export default class RemoteControlledSampler {
+
+    _sampler: Sampler;
+    _callerName: string;
+    _logger: any;
+    _timeoutHandle: any;
+    _intervalHandle: any;
+    _testCallback: Function;
+    _refreshInterval: number;
+
+    constructor(
+        callerName: string,
+        options: any
+        ) {
+
+        this._callerName = callerName;
+        this._sampler = options.sampler || new ProbabilisticSampler(DEFAULT_INITIAL_SAMPLING_RATE);
+        this._logger = options.logger || new NullLogger();
+        this._refreshInterval = options.refreshInterval || SAMPLING_REFRESH_INTERVAL;
+
+        let randomDelay: number = Math.random() * this._refreshInterval;
+        this._testCallback = options.testCallback;
+
+        // This branch is necessary because code in 'setTimeout' does not always run in the same tick
+        // thus in testing the code proceeds without calling 'options.callback', and the tests end without
+        // hitting all the proper assertions.
+        if (options.firstRefreshDelay) {
+            this._timeoutHandle = setTimeout(() => {
+                this._intervalHandle = setInterval(this._refreshSamplingStrategy, this._refreshInterval);
+            }, randomDelay)
+
+        } else {
+            this._refreshSamplingStrategy();
```

Comment:
In the above it should be "firstRefreshDelay: true, refreshInterval: 0.001," that runs, but then you encounter a problem where the testCallback calls "done" multiple times.